server/server.go

```diff
@@ -355,6 +366,9 @@ func (bs *Server) sendBlocks(ctx context.Context, env *decision.Envelope) {
 	bs.counters.DataSent += uint64(dataSent)
 	bs.counterLk.Unlock()
 	bs.sentHistogram.Observe(float64(env.Message.Size()))
+	if bs.onDataSentListener != nil {
+		bs.onDataSentListener(env.Peer, uint64(dataSent), uint64(len(blocks)))
```

Comment:
Why just send a block total ? I feel this could be made much more generic just by passing `blocks` instead.
Right now it feels like a very adhoc solution to the immediate your are having instead of a thought through solution.

Since it's a slice already the performance cost is minimal (it will pass one more pointer and one more `int`).

Reply:
honestly, because of previous bad experience letting memory escape like that. By baseline rule with hooks is don't give people references to large pools of memory on the possibility they hold it somehow. It's a defensive programming based of previous experience with graphsync. Actually, I don't even need the blocks sent anyway so maybe it should just be the bytes.

Reply:
Fair enough, you can keep it as-is.